packages/lit-dev-content/src/components/litdev-playground-share-button.ts

```diff
@@ -13,10 +13,77 @@ import {LitElement, html, css} from 'lit';
 import {customElement, property, state, query} from 'lit/decorators.js';
 import {shareIcon} from '../icons/share-icon.js';
 
+import type {
+  PropertyValues,
+  ReactiveController,
+  ReactiveControllerHost,
+} from 'lit';
 import type {LitDevPlaygroundShareLongUrl} from './litdev-playground-share-long-url.js';
+import type {LitDevPlaygroundShareGist} from './litdev-playground-share-gist.js';
 import type {SampleFile} from 'playground-elements/shared/worker-api.js';
 import type {Snackbar} from '@material/mwc-snackbar';
 
+/**
+ * Handles the Ctrl+S and Cmd+S keyboard shortcuts.
+ *
+ * This is factored out for readability. Though it could also be nice to make it
+ * more generic by taking a map of key-sequences and callbacks.
+ */
+class KeyboardController implements ReactiveController {
+  private readonly _callback: () => void;
+  private _ctrlDown = false;
+  private _cmdDown = false;
+
+  constructor(host: ReactiveControllerHost, callback: () => void) {
+    host.addController(this);
+    this._callback = callback;
+  }
+
+  hostConnected() {
+    window.addEventListener('keydown', this._onKeydown);
+    window.addEventListener('keyup', this._onKeyup);
+    window.addEventListener('blur', this._reset);
+  }
+
+  hostDisconnected() {
+    window.removeEventListener('keydown', this._onKeydown);
+    window.removeEventListener('keyup', this._onKeyup);
+    window.removeEventListener('blur', this._reset);
+    this._reset();
+  }
+
+  private readonly _onKeydown = (event: KeyboardEvent) => {
+    if (event.key === 'Control') {
+      this._ctrlDown = true;
+    } else if (event.key === 'Meta') {
+      this._cmdDown = true;
+    } else if (
+      // TODO(aomarks) File a Playwright issue. Playwright seems to have a bug
+      // where the "key" property is always uppercase, even when Shift was not
+      // also held. This seems to violate the UI Events spec
+      // (https://www.w3.org/TR/uievents-key/#:~:text=the%20key%20attribute%20value%20for%20the,%22q%22).
+      event.key.toLowerCase() === 's' &&
```

Comment:
Interesting! Yes, but I didn't realize it did.

https://w3c.github.io/uievents/tools/key-event-viewer.html

Turns out if you type `S+Ctrl` you get an `S` event with `ctrlKey:false`, then a `Control` event, then an `S` event with `ctrlKey:true`. So the logic we had actually applied there too.

I'm thinking it's ok to work this way? VSCode doesn't treat `S+Ctrl` the same as `Ctrl+S`, but Chrome actually does -- you can type `S+Ctrl` to trigger the default save-page dialog, or `F+Ctrl` to open the find menu (at least in Linux where I'm testing). I've never accidentally done this or noticed it before, personally.

And since I was investigating this, I discovered the `ctrlKey` and `metaKey` properties, which are supported in all browsers (https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/ctrlKey). So it's actually way simpler to implement the keyboard shortcut by just checking that property -- we don't need to keep track of those key states ourselves at all.

I've therefore moved all the logic into one function in the share button and gotten rid of the controller.

I also realized that we should ignore `repeat` events, since otherwise if you type `Ctrl+S` and keep the key down long enough for your repeat to fire, it will keep saving rapid fire!